_LowLevelCode/cpp/cpp_communicator/serialiser/c_deserialise.cpp

```diff
@@ -0,0 +1,363 @@
+ * c_deserialise.cpp
+ * Deserialise serialised data back into a MATLAB object
+ *
+ * See also:
+ * hlp_serialize
+ * hlp_deserialize
+ *
+ * This is a MEX-file for MATLAB.
+ *=======================================================*/
+
+#include <iostream>
+#include <cstring>
+#include <cmath>
+#include <vector>
+#include "cpp_serialise.hpp"
+
+size_t memPtr;
+
+template<typename T>
+inline void deser(const uint8_t* data, std::vector<T> output, const double amount) {
+  memcpy(output.data(), &data[memPtr], amount);
+  memPtr += amount;
+}
+
+inline void deser(const uint8_t* data, void* output, const double amount) {
+  memcpy(output, &data[memPtr], amount);
+  memPtr += amount;
+}
+
+inline void read_data(uint8_t* data, mxArray* output, const size_t elemSize, const size_t nElem) {
+  if (mxIsComplex(output)) {
+    // Size of a complex component is half that of the whole double
+    size_t compSize = elemSize/2;
+
+#if MX_HAS_INTERLEAVED_COMPLEX
+    void* toWrite = mxGetComplexDouble(input);
+    // Offset imaginary to end
+    size_t imPtr = memPtr + (1+nElem)*compSize;
+
+    for (int i = 0; i < nElem; i+=elemSize, memPtr += compSize, imPtr += compSize) {
+      memcpy(&toWrite[i]         , &data[memPtr],  compSize);
+      memcpy(&toWrite[i+compSize], &data[imPtr] ,  compSize);
+    }
+    memPtr = imPtr; // Finally update memPtr
+
+#else
+    void* toWrite = mxGetPr(output);
+    deser(data, toWrite, compSize*nElem);
+    toWrite = mxGetPi(output);
+    deser(data, toWrite, compSize*nElem);
+
+#endif
+
+  } else {
+    void* toWrite = mxGetPr(output);
+    deser(data, toWrite, elemSize*nElem);
+  }
+}
+
+
+mxArray* deserialise(uint8_t* data, size_t size, bool recursed) {
+
+  mxArray* output = NULL;
+
+  tag_type tag;
+  deser(data, &tag, types_size[UINT8]);
+  size_t nDims;
+  std::vector<uint32_t> cast_dims(2);
+  std::vector<mwSize> vDims(2);
+  size_t nElem;
+
+  // Special case as function handles work differently
+  if (tag.type != FUNCTION_HANDLE) {
+    nDims = tag.dim;
+    deser(data, cast_dims, nDims*types_size[UINT32]);
+    switch (nDims) {
+    case 0:
+      nElem = 1;
+      nDims = 2;
+      std::fill(vDims.begin(), vDims.end(), 1);
+      break;
+
+    case 1:
+      nElem = cast_dims[0];
+      nDims = 2;
+      if (nElem == 0) { // Handle null
+        std::fill(vDims.begin(), vDims.end(), 0);
+      } else {
+        vDims[0] = 1;
+        vDims[1] = nElem;
+      }
+      break;
+
+    default:
+      if (nDims > 2) {
+        vDims.resize(nDims);
+        cast_dims.resize(nDims);
+      }
+      nElem = 1;
+
+      for (size_t i = 0; i < nDims; i++) {
+        vDims[i] = cast_dims[i];
+        nElem *= vDims[i];
+      }
+      break;
+
+    }
+  } else {
+    // Function handle always scalar
+    std::fill(vDims.begin(), vDims.end(), 1);
+    std::fill(cast_dims.begin(), cast_dims.end(), 1);
+    nElem = 1;
+
+  }
+
+  // Shorthand for passing to c function
+  mwSize* dims = vDims.data();
```

Comment:
Because the MEX C API requires `mwSize*`, rather than `std::vector<mwSize>`, rather than writing `vDims.data()` or `&vDims[0]` everywhere that's used, I just declare a shorthand variable.

Reply:
So a comment like "making ptr to vdims.data as std::vector cannot be passed through the MEX C API" would clarify what shorthand means.